react/hooks/useFacetNavigation.js

```diff
@@ -5,109 +5,177 @@ import { useQuery } from '../components/QueryContext'
-            ].join('/'),
+              ...query.slice(0, lastCategoryIndex + 1),
+              facetValue,
+              ...query.slice(lastCategoryIndex + 1),
+            ],
             map: [
-              ...mapArray.slice(0, lastCategoryIndex + 1),
+              ...map.slice(0, lastCategoryIndex + 1),
               facet.map,
-              ...mapArray.slice(lastCategoryIndex + 1),
-            ].join(','),
+              ...map.slice(lastCategoryIndex + 1),
+            ],
           }
         }
       }
 
       return {
-        query: `${query}/${facet.value}`,
-        map: `${map},${facet.map}`,
+        query: [...query, facetValue],
+        map: [...map, facet.map],
       }
     },
-    { query: inputQuery, map: inputMap }
-  )
-
-const useFacetNavigation = () => {
-  const { navigate, setQuery } = useRuntime()
-  const { query, map } = useQuery()
-
-  const navigateToFacet = useCallback(
-    (maybeFacets, preventRouteChange = false) => {
-      const facets = Array.isArray(maybeFacets) ? maybeFacets : [maybeFacets]
-
-      const { query: currentQuery, map: currentMap } = buildQueryAndMap(
-        query,
-        map,
-        facets
-      )
-
-      if (preventRouteChange) {
-        setQuery({
-          map: `${currentMap}`,
-          query: `/${currentQuery}`,
-          page: undefined,
-        })
-        return
-      }
-
-      const urlParams = new URLSearchParams(window.location.search)
-
-      urlParams.set('map', currentMap)
-      urlParams.delete('page')
-      urlParams.delete('query')
-
-      navigate({
-        to: `/${currentQuery}`,
-        query: urlParams.toString(),
-        scrollOptions,
-      })
-    },
-    [query, map, navigate, setQuery]
+    { query: querySegments, map: mapSegments }
   )
+  return {
+    query: queryAndMap.query.join(PATH_SEPARATOR),
+    map: removeMapForNewURLFormat(queryAndMap).join(MAP_VALUES_SEP),
+  }
+}
 
+const useFacetNavigation = map => {
+  const { navigate } = useRuntime()
+  const { query } = useQuery()
+
+  const navigateToFacet = useCallback(maybeFacets => {
+    const facets = Array.isArray(maybeFacets) ? maybeFacets : [maybeFacets]
+
+    const querySegments = query.split(PATH_SEPARATOR)
+    const mapSegments = map.split(MAP_VALUES_SEP)
+
+    const {
+      query: newQuerySegments,
+      map: newMapSegments,
+    } = convertSegmentsToNewURLs(querySegments, mapSegments, selectedFacets)
+
+    const { query: currentQuery, map: currentMap } = buildQueryAndMap(
+      newQuerySegments,
+      newMapSegments,
+      facets
+    )
+
+    const urlParams = getCleanUrlParams(currentMap)
+    const modifiersIgnore = getStaticPathSegments([
+      ...Object.values(selectedFacets),
+      ...facets,
+    ])
+
+    navigate({
+      to: `${PATH_SEPARATOR}${currentQuery}`,
+      query: urlParams.toString(),
+      scrollOptions,
+      modifiersIgnore,
+    })
+  })
+
+  navigateToFacet.storeSelectedFacets = storeSelectedFacets
```

Comment:
I get what you did but modifying the object of a `useCallback` call is completely against React pattern. You can make this hook return a tuple [navigateToFacet, storeSelectedFacets], something like that, but do change navigateToFacet like this